src/test/gather_node_test.cpp

```diff
@@ -50,35 +50,123 @@ class GatherNodeInputHandlerTest : public ::testing::Test {};
+            ASSERT_FALSE(gInputHandler.isReady());
+            SPDLOG_ERROR("i: {}, j: {}, metadatas.size: {}, metadatas[i].size() :{}", i, j, metadatas.size(), metadatas[i].size());
+            auto shardId = metadatas[i][j].getShardId({demultiplexerNodeNames[0], demultiplexerNodeNames[1]});
+            status = gInputHandler.setInput(inputName,
+                blob,
+                shardId);
+            ASSERT_EQ(status.getCode(), StatusCode::OK) << status.string();
+            gInputHandler.notifyFinishedDependency();
+        }
+    }
+    ASSERT_TRUE(gInputHandler.isReady());
+    const auto blobMap = gInputHandler.getInputs();
+    ASSERT_EQ(blobMap.size(), 1);
+    const auto& blob = blobMap.at(inputName);
+    EXPECT_EQ(blob->size(), blobsData.size());
+    EXPECT_THAT(blob->getTensorDesc().getDims(), ElementsAre(1, demultiplyCounts[0], demultiplyCounts[1], elementCountPerShard));
+    EXPECT_EQ(std::memcmp((char*)((const void*)blob->cbuffer()), blobsData.data(), blobsData.size() * sizeof(float)), 0);
+}
+
+TEST_F(GatherNodeInputHandlerTest, SetInputsWithShardsHavingDifferentShapesShouldReturnErrorWhenGathering) {
+    // simulate all 3 inputs comming from different predecessor nodes
+    // with session demultiplexed to 2 shards
+    const std::string inputNames{"a"};
+    std::vector<std::vector<size_t>> shapes{{1, 10}, {1, 9}};
+    InferenceEngine::Precision precision{InferenceEngine::Precision::FP32};
+    InferenceEngine::Layout layout{InferenceEngine::Layout::NC};
+    std::vector<float> blobsData{-1, 4, 5, 12, 3, 52, 12, 0.5, 9, 1.67};
+    std::vector<InferenceEngine::TensorDesc> descs{{precision, shapes[0], layout}, {precision, shapes[1], layout}};
+    std::vector<InferenceEngine::Blob::Ptr> inputBlobs{InferenceEngine::make_shared_blob<float>(descs[0], blobsData.data()), InferenceEngine::make_shared_blob<float>(descs[1], blobsData.data())};
+    const session_id_t shardsCount = 2;  // subsessionSize/demultiplyCount
+    CollapsingDetails collapsingDetails{{std::string("NOT_IMPORTANT_DEMULTIPLEXER_NAME")}, {shardsCount}};
+    GatherNodeInputHandler gInputHandler(inputNames.size(), collapsingDetails);
+    Status status;
+    for (session_id_t j = 0; j < shardsCount; ++j) {
+        EXPECT_FALSE(gInputHandler.isReady());
+        status = gInputHandler.setInput(inputNames, inputBlobs[j], j);
+        // each input comming from different node so we call notify each time
+        if (!status.ok()) {
+            SPDLOG_ERROR("ER");
+            EXPECT_EQ(status.getCode(), StatusCode::PIPELINE_INCONSISTENT_SHARD_DIMENSIONS) << status.string();
+            break;
+        }
+        status = gInputHandler.notifyFinishedDependency();
+        EXPECT_EQ(status.getCode(), StatusCode::OK) << status.string();
+    }
+    // The second notify should fail since the shard dimension should be different
+    EXPECT_FALSE(status.ok());
```

Comment:
Expect specific error status here?

Reply:
2 lines below - but I should have simpliefied that anyway.